File: TigerConnect Demo/Views/OrganizationsBarButtonView.m
//
//  OrganizationsBarButtonView.m
//  TigerConnect Demo
//
//  Created by Oren Zitoun on 2/2/16.
//  Copyright © 2016 TigerText. All rights reserved.
//

#import "OrganizationsBarButtonView.h"
#import "UIHelpers.h"

#import <TTKit/TTKit.h>

@interface OrganizationsBarButtonView ()

@property (nonatomic) UIImageView *organizationsImageView;
@property (nonatomic) UIView *unreadMessagesBadge;

@end

@implementation OrganizationsBarButtonView

- (void)dealloc
{
    [[NSNotificationCenter defaultCenter] removeObserver:self];
}

+ (instancetype)barView
{
    OrganizationsBarButtonView *view = [[OrganizationsBarButtonView alloc] initWithFrame:CGRectMake(0, 0, 26, 25)];
    return view;
}

- (instancetype)initWithFrame:(CGRect)frame
{
    self = [super initWithFrame:frame];
    if (self) {
        self.organizationsImageView = [[UIImageView alloc] initWithFrame:CGRectMake(0, 0, 22, 22)];
        self.organizationsImageView.left = 0;
        self.organizationsImageView.bottom = CGRectGetHeight(frame);
        self.organizationsImageView.image = [[UIImage imageNamed:@"orgs"] imageWithRenderingMode:UIImageRenderingModeAlwaysTemplate];
        self.organizationsImageView.tintColor = [UIColor whiteColor];
        [self addSubview:self.organizationsImageView];
        
        self.unreadMessagesBadge = [[UIView alloc] initWithFrame:CGRectMake(0, 0, 8, 8)];
        self.unreadMessagesBadge.backgroundColor = [UIColor colorFromRGB:0xff3838];
        self.unreadMessagesBadge.cornerRadius = 4.0f;
        self.unreadMessagesBadge.borderColor = [UIColor whiteColor];
        self.unreadMessagesBadge.borderWidth = 1.0f;
        self.unreadMessagesBadge.right = CGRectGetMaxX(frame);
        self.unreadMessagesBadge.top = 0;
        [self addSubview:self.unreadMessagesBadge];
        [self refreshBadge];
        [self registerNotifications];
        
        UIButton *overlayButton = [[UIButton alloc] initWithFrame:CGRectMake(0, 0, CGRectGetWidth(frame), CGRectGetHeight(frame))];
        overlayButton.backgroundColor = [UIColor clearColor];
        [overlayButton addTarget:self action:@selector(onOverlayButtonPressed:) forControlEvents:UIControlEventTouchUpInside];
        [self addSubview:overlayButton];
    }
    return self;
}

- (void)registerNotifications
{
    [[NSNotificationCenter defaultCenter] addObserver:self
                                             selector:@selector(refreshBadge)
                                                 name:kTTKitUnreadMessagesCountChangedNotification
                                               object:nil];
    
    [[NSNotificationCenter defaultCenter] addObserver:self
                                             selector:@selector(refreshBadge)
                                                 name:kTTKitCurrentOrganizationChangeNotification
                                               object:nil];
}

- (void)refreshBadge
{
    NSUInteger unreadCount = 0;
    NSString *currentOrgToken = [[TTKit sharedInstance] currentOrganizationToken];
    if(currentOrgToken) {
        TTBadgeData *orgBadgeData = [[TTKit sharedInstance] badgeDataForOrganizationToken:currentOrgToken];
        unreadCount = orgBadgeData.unreadCount;
    }
    
    NSUInteger totalUnreadCount = [[TTKit sharedInstance] getTotalUnreadMessageCount];
    if(totalUnreadCount - unreadCount > 0) {
        self.unreadMessagesBadge.hidden = NO;
    } else {
        self.unreadMessagesBadge.hidden = YES;
    }
}

- (void)onOverlayButtonPressed:(id)sender
{
    if ([self.delegate respondsToSelector:@selector(organizationBarButtonPressed:)]) {
        [self.delegate organizationBarButtonPressed:self];
    }
}

@end

File: TC Swift Demo/Views/OrganizationsBarButtonView.swift
//
//  OrganizationsBarButtonView.swift
//  TC Swift Demo
//
//  Created by Oren Zitoun on 2/12/16.
//  Copyright © 2016 TigerConnect. All rights reserved.
//

import Foundation

protocol OrganizationsBarButtonViewDelegate {
    func organizationBarButtonPressed(_ sender: OrganizationsBarButtonView!)
}

class OrganizationsBarButtonView: UIView {
    fileprivate var organizationsImageView: UIImageView!
    fileprivate var unreadMessagesBadge: UIView!
    var delegate: OrganizationsBarButtonViewDelegate?
    
    deinit {
        NotificationCenter.default.removeObserver(self)
    }
    
    class func barView() -> OrganizationsBarButtonView {
        let view = OrganizationsBarButtonView.init(frame: CGRect(x: 0, y: 0, width: 26, height: 25))
        return view
    }
    
    required init?(coder aDecoder: NSCoder) {
        super.init(coder: aDecoder)
    }
    
    override init(frame:CGRect) {
        super.init(frame:frame)
        self.organizationsImageView = UIImageView.init(frame: CGRect(x: 0, y: 0, width: 22, height: 22))
        self.organizationsImageView.left = 0
        self.organizationsImageView.bottom = frame.height
        self.organizationsImageView.image = UIImage.init(named: "orgs")?.withRenderingMode(UIImageRenderingMode.alwaysTemplate)
        self.organizationsImageView.tintColor = UIColor.white
        self.addSubview(self.organizationsImageView)
        
        self.unreadMessagesBadge = UIView.init(frame: CGRect(x: 0, y: 0, width: 8, height: 8))
        self.unreadMessagesBadge.backgroundColor = UIColor.init(hex: 0xff3838)
        self.unreadMessagesBadge.cornerRadius = 4.0
        self.unreadMessagesBadge.borderColor = UIColor.white
        self.unreadMessagesBadge.borderWidth = 1.0
        self.unreadMessagesBadge.right = frame.maxX
        self.unreadMessagesBadge.top = 0
        self.addSubview(self.unreadMessagesBadge)
        self.refreshBadge()
        self.registerNotifications()
        
        let overlayButton = UIButton.init(frame: CGRect(x: 0, y: 0, width: frame.width, height: frame.height))
        overlayButton.backgroundColor = UIColor.clear
        overlayButton.addTarget(self, action: #selector(OrganizationsBarButtonView.onOverlayButtonPressed), for: UIControlEvents.touchUpInside)
        self.addSubview(overlayButton)
    }
    
    fileprivate func registerNotifications() {
        NotificationCenter.default.addObserver(self, selector: #selector(OrganizationsBarButtonView.refreshBadge), name: NSNotification.Name.ttKitUnreadMessagesCountChanged, object: nil)
        NotificationCenter.default.addObserver(self, selector: #selector(OrganizationsBarButtonView.refreshBadge), name: NSNotification.Name.ttKitCurrentOrganizationChange, object: nil)
    }
    
    @objc fileprivate func refreshBadge() {
        var unreadCount: UInt = 0
        let currentOrgToken = TTKit.sharedInstance().currentOrganizationToken()
        if (currentOrgToken != nil) {
            let badgeData = TTKit.sharedInstance().badgeData(forOrganizationToken: currentOrgToken)
            unreadCount = (badgeData?.unreadCount)!
        }
        
        let totalUnreadCount = TTKit.sharedInstance().getTotalUnreadMessageCount()
        if (UInt(totalUnreadCount - unreadCount)) > 0 {
            self.unreadMessagesBadge.isHidden = false
        } else {
            self.unreadMessagesBadge.isHidden = true
        }
    }
    
    func onOverlayButtonPressed() {
        if let delegate = self.delegate {
            delegate.organizationBarButtonPressed(self)
        }
    }
    
}
